### PR TITLE
SGDR T_mult=2 with best-checkpoint saving across all 4 CFD benchmarks

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

SGDR with `T_mult=2` (Cosine Annealing with Warm Restarts where each cycle is twice as long as the previous) produces progressively longer exploration-exploitation cycles. The first restart is aggressive, the later ones increasingly conservative — a natural curriculum. The key question is whether the model hits transient minima at restart points that are better than where the long cosine run ends.

This hypothesis was attempted previously but lacked best-checkpoint saving, so transient minima at restart points were never captured. PR #3029 (megumi) introduced `checkpoint_best.pt` saved whenever val improves. With that infrastructure now in place, SGDR T_mult deserves a clean retry where we actually capture the best checkpoint across all restarts rather than only evaluating the terminal epoch.

**Cross-dataset scope:** TandemFoil · TandemFoil Paper · AirfRANS · DrivAerML

---

## Instructions to the Student

Use `torch.optim.lr_scheduler.CosineAnnealingWarmRestarts` with `T_0` matching the dataset T_max baseline and `T_mult=2`. This gives restart intervals of T_0, 2*T_0, 4*T_0, ... epochs.

**Critical:** The best-checkpoint mechanism from #3029 must be active so transient minima at each restart are captured. Confirm `checkpoint_best.pt` is being saved (it should be automatic in the current codebase). Report the best val metric from the best checkpoint, not the terminal epoch.

Dataset-specific T_0 values (matching baseline cosine-t-max):
- TandemFoil: T_0=10
- TandemFoil Paper: T_0=150
- AirfRANS: T_0=50
- DrivAerML: T_0=30

Use `--wandb_group sgdr-tmult-best-checkpoint` on all runs.

---

### TandemFoil

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --cosine-t-mult 2 \
  --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame \
  --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999 \
  --wandb_group sgdr-tmult-best-checkpoint
```

(If `--cosine-t-mult` is not yet a flag, implement it: add `T_mult` parameter to `CosineAnnealingWarmRestarts`; if only `CosineAnnealingLR` is used, switch to `CosineAnnealingWarmRestarts` with `T_0=cosine_t_max, T_mult=2`.)

### TandemFoil Paper

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 150 --cosine-t-mult 2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --epochs 999 \
  --wandb_group sgdr-tmult-best-checkpoint
```

### AirfRANS

```bash
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 \
  --cosine-t-max 50 --cosine-t-mult 2 \
  --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group sgdr-tmult-best-checkpoint
```

### DrivAerML

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 \
  --cosine-t-max 30 --cosine-t-mult 2 \
  --no-use-ema \
  --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 \
  --batch-size 1 \
  --drivaerml-train-surface-points 50000 \
  --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --wandb_group sgdr-tmult-best-checkpoint
```

---

## Current Baselines (lower is better)

| Dataset | Metric | Current Best | PR |
|---|---|---|---|
| TandemFoil | `val_primary/surface_pressure_mae` | 22.537 | #2924 |
| TandemFoil Paper | `val_primary/field_mse` | 0.00434 | #2979 |
| AirfRANS | `val_primary/surface_mse` | 0.000482 | #2951 |
| DrivAerML | `val_primary/surface_rel_l2_pct` | 3.997% | #2898 |

---

## Reporting

When done, post results as a PR comment with:
- W&B run IDs for all runs
- Best val metric per dataset from `checkpoint_best.pt` (not terminal epoch)
- Whether SGDR T_mult=2 beat baseline on each dataset
- At which restart interval the best checkpoint was captured
- Comparison with terminal epoch metrics (to quantify transient minima benefit)

### CODE REVIEW

**Comment by morganmcg1** [comment]
Closing to narrow the queue for the final ICML benchmark sprint. Current priority is benchmark-specific optimization and benchmark-faithful reporting for DrivAerML, TandemFoil paper benchmark, AirfRANS full Surf/Vol, and the Tandem parity anchor. If this idea still matters later, we can reopen it in a benchmark-specific form.
